src/iter_index.rs

```diff
@@ -0,0 +1,114 @@
+use core::iter::{Skip, Take};
+use core::ops::{Range, RangeFrom, RangeFull, RangeInclusive, RangeTo, RangeToInclusive};
+
+#[cfg(doc)]
+use crate::Itertools;
+
+mod private_iter_index {
+    use core::ops;
+
+    pub trait Sealed {}
+
+    impl Sealed for ops::Range<usize> {}
+    impl Sealed for ops::RangeInclusive<usize> {}
+    impl Sealed for ops::RangeTo<usize> {}
+    impl Sealed for ops::RangeToInclusive<usize> {}
+    impl Sealed for ops::RangeFrom<usize> {}
+    impl Sealed for ops::RangeFull {}
+}
+
+/// Used by [`get`] and [`Itertools::get`] to know which iterator
+/// to turn different ranges into.
+pub trait IteratorIndex<I>: private_iter_index::Sealed
+where
+    I: Iterator,
+{
+    /// The type returned for this type of index.
+    type Output: Iterator<Item = I::Item>;
+
+    /// Returns an adapted iterator for the current index.
+    ///
+    /// Prefer calling [`get`] or [`Itertools::get`] instead
+    /// of calling this directly.
+    fn index(self, from: I) -> Self::Output;
+}
+
+impl<I> IteratorIndex<I> for Range<usize>
+where
+    I: Iterator,
+{
+    type Output = Take<Skip<I>>;
+
+    fn index(self, iter: I) -> Self::Output {
+        iter.skip(self.start)
+            .take(self.end.saturating_sub(self.start))
+    }
+}
+
+impl<I> IteratorIndex<I> for RangeInclusive<usize>
+where
+    I: Iterator,
+{
+    type Output = Take<Skip<I>>;
+
+    fn index(self, iter: I) -> Self::Output {
+        debug_assert!(!self.is_empty(), "The given `RangeInclusive` is exhausted. The result of indexing with an exhausted `RangeInclusive` is unspecified.");
```

Comment:
**Feedback from CI:** `RangeInclusive::is_empty` requires Rust 1.47.0+ so the MSRV check fails.
It can't be inlined because of a private field (`exhausted`). Our options:
- Wait that the MSRV hits 1.47.0 to merge this.
- (Merge this feature without `RangeInclusive`, it would be really weird for the user).
- Or temporarily disable this `debug_assert`:
```suggestion
        // TODO: Once MSRV >= 1.47.0, uncomment:
        // debug_assert!(!self.is_empty(), "The given `RangeInclusive` is exhausted. The result of indexing with an exhausted `RangeInclusive` is unspecified.");
```

Reply:
@jswrenn What do you think we should do here?

Reply:
So long as we document that the result of indexing with an exhausted `RangeInclusive` is unspecified, we can remove the `debug_assert!`.

Reply:
The documentation of `Itertools::get` mentions it so I'm gonna remove this.